feat(TestScheduler): add an animate "run mode" helper

cartant opened the pull request:
<!--
Thank you very much for your pull request!

If your PR is the addition of a new operator, please make sure all these boxes are ticked with an x:

- [ ] Add the operator to Rx
- [ ] It must have a `-spec.ts` tests file covering the canonical corner cases, with marble diagram tests
- [ ] The spec file should have a type definition test at the end of the spec to verify type definition for various use cases
- [ ] The operator must be documented in JSDoc style in the implementation file, including also the PNG marble diagram image
- [ ] The operator should be listed in `docs_app/content/guide/operators.md` in a category of operators
- [ ] The operator should also be documented. See [Documentation Guidelines](../CONTRIBUTING.md).
- [ ] You may need to update `MIGRATION.md` if the operator differs from the corresponding one in RxJS v4
-->

**Description:**

This PR:

* Adds a `Date` timestamp provider.
* Adds a `performance` timestamp provider.
* Adds a `requestAnimationFrame` provider.
* All providers have `delegate` properties that are `undefined` by default but are used - by `TestScheduler#run` - to override the default behaviour. This mirrors the `delegate` implementation detail that's used with the schedulers.
* Removes the arrow function from the static `Scheduler.now` - it's unnecessary, as the `Date.now` call now occurs within the `dateTimestampProvider`.
* The `animationFrames` observable now uses the `requestAnimationFramesProvider`.
* `TestScheduler#run` sets the delegates for the providers, with the `requestAnimationFramesProvider`'s delegate working in conjunction with a `animate()` marble that's added to the run helpers:

    ```ts
    testScheduler.run(({ animate }) => {
      animate('---x---x---x---x---x');
      /* ... */
    });
    ```

    Each time an animation/repaint occurs, all queued `rAF` callbacks will be executed - in the order in which they were scheduled - and each will be passed the same timestamp - the `TestScheduler`'s `now` value - which is also what will be returned from the timestamp providers - as the `TestScheduler` itself is the delegate.

* Adds tests for the `animate` `run` helper.
* Updates API guardian for the inclusion of the `animate` `run` helper.
* Reimplements the `animationFrames` tests to use the `run` helper and the provider - which is spied upon - and removes the `rAF` mock.
* Fixes the `rAF` provider so that cancellation of already fulfilled requests (upon unsubscription) is not attempted.

**Related issue (if exists):** #5438 and #5475

benlesh commented on api_guard/dist/types/index.d.ts:
```diff
@@ -618,7 +618,7 @@ export declare class VirtualAction<T> extends AsyncAction<T> {
     protected requestAsyncId(scheduler: VirtualTimeScheduler, id?: any, delay?: number): any;
     schedule(state?: T, delay?: number): Subscription;
-    static sortActions<T>(a: VirtualAction<T>, b: VirtualAction<T>): 1 | 0 | -1;
+    static sortActions<T>(a: VirtualAction<T>, b: VirtualAction<T>): 0 | 1 | -1;
```
LOL.

benlesh commented on src/internal/testing/TestScheduler.ts:
```diff
@@ -409,14 +413,62 @@ export class TestScheduler extends VirtualTimeScheduler {
-    const helpers = {
+    let animationFramesHandle = 0;
+    let animationFramesQueue: Map<number, FrameRequestCallback> | undefined;
+
```
Be sure to add comments as to the "why" around here. We don't do that enough in our library, I think, and while it might be obvious to us right now, maybe it won't be 5 years from now.

benlesh commented on src/internal/testing/TestScheduler.ts:
```diff
@@ -409,14 +413,62 @@ export class TestScheduler extends VirtualTimeScheduler {
+        animationFramesQueue.delete(handle);
+      }
+    };
+    const repaints = (marbles: string) => {
```
I can easily have my mind changed on this... but I think we should probably move this out to have it in the same place as all the other helpers. I know that makes it available for "non-run-mode" use, and maybe we want to avoid that.. but I do like consistency.

benlesh replied:
Oh... on more thing, I'm not sure I like `repaints`, as it's not readily obvious that it is associated to "animation frames". Perhaps a different name here would be better?  `animates`? `animationFrames`? (the latter might be too confusing though).

cartant replied:
> Perhaps a different name here would be better? 

How about `animate`? It doesn't return anything, so I suppose it doesn't need the plural noun that would usually be used for an observable.

benlesh commented on src/internal/testing/TestScheduler.ts:
```diff
@@ -409,14 +413,62 @@ export class TestScheduler extends VirtualTimeScheduler {
+        animationFramesQueue.delete(handle);
+      }
+    };
+    const repaints = (marbles: string) => {
```
this also is going to need documentation comments.

cartant commented:
@benlesh I'll make those changes and will then add some docs - I was holding off on documenting it until it had been reviewed.

cartant commented:
@benlesh I've made some changes:

* I've changed the name of the `repaints` run helper to `animate`.
* I've moved the delegate and the run helper implementation into a `createAnimator` method. This is more consistent with the current implementation.
* I've added a TSDoc comment to `run` - there was no existing comment/documentation - and I've linked it to an updated marble testing guide.

ATM, the documentation in the guide is a little light, but I think it's enough, for now. Once this PR is merged, I am going to:

* rebase and fix up the PR that adds the `timestamp` and `elapsed` properties to the `animationFrames` observable's value.
* look at removing/refactoring the `delegate` that's in `AsyncScheduler`, as I think to ought to be possible to add `setInterval` and `setImmediate` providers (and delegates). Doing that would allow for tests to actually use all of the schedulers within 'run mode'. ATM, they are bypassed and there is no way to write a test in which `AsyncScheduler` tasks are executed *after* subsequently-scheduled `AsapScheduler` tasks - i.e. there is no differentiation between virtual micro and macro tasks in the `TestScheduler`.

If the docs need more work, I can look at them after I've done the above. Or someone else can do it.

benlesh commented:
Nice work, @cartant! I'm very pleased to see this land!
